man/tilestache-clean.1
.TH TILESTACHE-CLEAN 1 "Sep 23, 2012"
.SH NAME
tilestache\-clean \- clean tiles from a TileStache cache
.SH SYNOPSIS
.B tilestache\-clean
.RI [ options ] " zoom" ...
.SH DESCRIPTION
This manual page documents briefly the \fBtilestache\-clean\fR command.
.PP
\fBtilestache\-clean\fP is used to clean tiles from a TileStache cache.
.PP
Cleans a single layer in your TileStache configuration - no images are returned,
and TileStache ends up with an empty in selected areas cache. Bounding box is
given as a pair of lat/lon coordinates, e.g. "37.788 \-122.349 37.833 \-122.246".
Output is a list of tile paths as they are created.
.PP
The \fIzoom\fR option is a space separated list of zoom levels which will be
cleaned. e.g. 12 13 14 15 will clean zooms 12 through 15. Many shells allow you
to use {a..b} which will expand to a list of numbers a though to b.
\fBRequired.\fR
.SH REQUIRED OPTIONS
.TP
\fB-c\fR, \fB\-\-config\fR \fIfile\fR
Path to configuration file. \fBRequired.\fR
.TP
\fB-l\fR, \fB\-\-layer\fR \fIlayer\fR
Layer name from configuration. "ALL" is a special value that will clean all
layers in turn. If you have an actual layer named "ALL", use "ALL LAYERS"
instead. \fBRequired.\fR
.TP
\fB-b\fR, \fB\-\-bbox\fR \fIsouth west north east\fR
Bounding box in floating point geographic coordinates. \fBRequired.\fR
.SH OPTIONS
.B \-h, \-\-help
Show summary of options.
.TP
\fB-p\fR, \fB\-\-padding\fR \fIpadding\fR
Extra margin of tiles to add around bounded area.
Default value is 0 (no extra tiles).
.TP
\fB-e\fR, \fB\-\-extension\fR \fIextension\fR
Optional file type for rendered tiles. Default value is 'png'.
.TP
\fB-f\fR, \fB\-\-progress\-file\fR \fIfile\fR
Optional JSON progress file that gets written on each iteration, so you don't
have to pay close attention.
.TP
\fB-q\fR
Suppress chatty output, \-\-progress\-file works well with this.
.TP
\fB-i\fR, \fB\-\-include\-path\fR \fIpath\fR
Add the following colon-separated list of paths to Python's include path (aka
sys.path)
.TP
\fB\-\-tile\-list\fR \fIfile\fR
Optional file of tile coordinates, a simple text list of Z/X/Y coordinates.
Overrides \-\-bbox and \-\-padding.
.SH SEE ALSO
.BR tilestache-seed (1)
.SH AUTHOR
\fBTileStache\fR was written by Michal Migurski <mike@stamen.com>.
.PP
This manual page was written by Andrew Harvey <andrew.harvey4@gmail.com>,
for the Debian project (and may be used by others).



CHANGELOG
2012-09-18: 1.42.0
- Added support for tiled bitmap files in Sandwich provider layer src attribute.

2012-09-14: 1.41.0
- Added TileStache.Goodies.Providers.Cascadenik to skip Mapnik XML where possible.

2012-09-09: 1.40.1
- Re-attempting Vector import to trigger a more-useful error than NameError in getProviderByName().

2012-09-04: 1.40.0
- Added TileStache.Sandwich provider based on previous Composite provider under Goodies.
- Modified import method for provider to reduce burden of required packages and improve error output.

2012-07-18: 1.39.0
- Fixed missing merge_grids() function in Mapnik Grid provider.
- Added TileStache.Goodies.ExternalConfigServer from Ryan Breen.

2012-07-05: 1.38.0
- Added Mapnik Grid provider "layer id key" parameter.
- Grid internals: response encodings to explicit UTF-8 for Chromium and using mapnik.render_layer().

2012-07-05: 1.37.0
- Added TileStache.Goodies.Proj4Projection for arbitrary PROJ.4 support.
- Added per-layer "tile height" option, for non-256x256 tiles.

2012-06-27: 1.36.1
- Added more Memcache disconnect_all() calls.

2012-05-22: 1.36.0
- Added --jsonp-callback option to tilestache-seed.py.

2012-05-22: 1.35.1
- Updated tilestache-seed.py manpage to reflect new options.

2012-05-22: 1.35.0
- Added direct-to-S3 cache option to tilestache-seed.py script. 
- Changed tilestache-seed.py default extension behavior to cover JSON.

2012-05-21: 1.34.0
- Added TileStache.Goodies.StatusServer alternate WSGI server for redis-backed internal status checks.
- Added maskband parameter to TileStache.Goodies.Providers.GDAL provider.

2012-05-09: 1.33.1
- Fixed mapnik.Envelope/Box2d back-compatibility.

2012-05-09: 1.33.0
- Added Vector provider support for Spatialite, MySQL and Oracle from Andrew Semprebon.

2012-05-08: 1.32.3
- Fixed a dumb import error in TileStache.Providers.

2012-05-07: 1.32.2
- Disambiguated null vs. empty list behavior of Mapnik Grid fields. 

2012-05-06: 1.32.1
- Fixed output of Mapnik Grid to be pure UTF-8 instead of ASCII-encoded for size.

2012-05-06: 1.32.0
- Added core Mapnik Grid provider based on providers under TileStache.Goodies with contributions from Alexander Clausen, Tom Nightingale, Dave Leaver and Paul Smith.

2012-04-09: 1.31.1
- Fixed broken MBTiles support.

2012-04-07: 1.31.0
- Added support for per-extension redirects in Layer configuration.
- Added new TileStache.Goodies.Provider.UtfGridComposite provider.
- Added new TileStache.Goodies.Cache.GoogleCloud cache.
- Added support for buffers around features in TileStache.Goodies.Provider.MapnikGrid.

2012-04-05: 1.30.0
- Added Goodies.AreaServer for serving static, non-tile images.
- Added TileStache.Goodies.Providers.Monkeycache for serving tiles out of caches.
- Documented JSONP callbacks in TileStache.requestHandler().

2012-03-27: 1.29.3
- Fixed a NameError in tilestache-seed.py.

2012-03-18: 1.29.2
- Ensured that in-process recent tiles features works with metatiles.

2012-03-15: 1.29.1
- Added debug logging to Mapnik provider.

2012-03-14: 1.29.0
- Added configurable HTTP referer header to URL Template provider.
- Added optional precision parameter to Vector provider for JSON output.
- Added basic logging facility, including getTile() logging at DEBUG level.

2012-02-26: 1.28.0
- Added new ways to define layer boundaries, including lists and partial limits.

2012-02-26: 1.27.3
- Fixed a forehead slapper in tilestache-seed.

2012-02-26: 1.27.2
- Fixed Mapnik fonts path bug.

2012-02-20: 1.27.1
- Fixed WSGITileServer issue in new Configuration.index support.

2012-02-20: 1.27.0
- Added new Configuration.index support for configurable index page responses.
- Cleaned up some unnecessary errors  in preview and callback handling.

2012-02-17: 1.26.4
- Rewrote internals of tilestache-seed to support the no-config use case.

2012-02-16: 1.26.3
- Tweaked Vector mime-types and mapnik import details.

2012-02-15: 1.26.2
- Fixed os.rmdir() error handling in Caches.Disk.

2012-02-15: 1.26.1
- Changed lock in Providers.Mapnik to be a single global lock for the entire process instead of per-layer.

2012-02-13: 1.26.0
- Modified internals of Providers.Mapnik to ensure single-threaded access to mapnik.Map object.
- Fixed MapnikGrid so its documentation can be made even though no one has mapnik2.
- Added -i/--include-path flag to tilestache-compose.py script.

2012-02-06: 1.25.1
- Fixed tilestache-clean.py script to not complain on unsupported -e flag.

2012-02-01: 1.25.0
- Added "ALL" ("ALL LAYERS") option to tilestache-clean.py script.

2012-01-30: 1.24.1
- Fixed an error in WSGITileServer._response().

2012-01-25: 1.24.0
- Added support for HTTP Expires and Cache-Control headers with per-layer maximum cache age.
- Merged Per Liedman's fox for disk cache lock directory removal bug.

2012-01-23: 1.23.4
- Fixed a bug in Multi Cache provider remove().

2012-01-12: 1.23.3
- Changed import delay from 1.23.2 to a specific conditional around sqlite3 for Heroku.

2012-01-11: 1.23.2
- Delayed import of MBTiles and Vector providers until last minute.

2012-01-10: 1.23.1
- Added Nikolai's httplib patch in Proxy provider.

2012-01-10: 1.23.0
- Added new tilestache-clean.py script and required remove() method for caches.

2012-01-01: 1.22.0
- Fixed inconsistencies with cache-ignore setting when using tilestache-compose.py.
- Fixed use of write-cache setting in TileStache.Layer.
- Adding configuration file's directory to sys.path in Config.buildConfiguration().
- Stopped adding current working directory to sys.path for new classes.
- Added new TileStache.Goodies.Providers.MapnikGrid:Provider by github user danzel.

2011-12-25: 1.21.3
- Fixed a bug where I'm an idiot.

2011-12-25: 1.21.2
- Fixed a locking bug in tilestache-compose.py.

2011-12-25: 1.21.1
- Added current working directory to sys.path for new classes.

2011-12-18: 1.21.0
- Added new tilestache-compose.py script.
- Fixed preview body style to account for iOS viewport.

2011-12-12: 1.20.1

scripts/tilestache-clean.py
#!/usr/bin/env python
"""tilestache-clean.py will flush your cache.

This script is intended to be run directly. This example cleans the area around
West Oakland (http://sta.mn/ck) in the "osm" layer, for zoom levels 12-15:

    tilestache-clean.py -c ./config.json -l osm -b 37.79 -122.35 37.83 -122.25 -e png 12 13 14 15

See `tilestache-clean.py --help` for more information.
"""

from sys import stderr, path
from optparse import OptionParser

try:
    from json import dump as json_dump
except ImportError:
    from simplejson import dump as json_dump

#
# Most imports can be found below, after the --include-path option is known.
#

parser = OptionParser(usage="""%prog [options] [zoom...]

Cleans a single layer in your TileStache configuration - no images are returned,
and TileStache ends up with an empty in selected areas cache. Bounding box is
given as a pair of lat/lon coordinates, e.g. "37.788 -122.349 37.833 -122.246".
Output is a list of tile paths as they are created.

Configuration, bbox, and layer options are required; see `%prog --help` for info.""")

defaults = dict(extension='png', padding=0, verbose=True, bbox=(37.777, -122.352, 37.839, -122.226))

parser.set_defaults(**defaults)

parser.add_option('-c', '--config', dest='config',
                  help='Path to configuration file.')

parser.add_option('-l', '--layer', dest='layer',
                  help='Layer name from configuration. "ALL" is a special value that will clean all layers in turn. If you have an actual layer named "ALL", use "ALL LAYERS" instead.')

parser.add_option('-b', '--bbox', dest='bbox',
                  help='Bounding box in floating point geographic coordinates: south west north east.',
                  type='float', nargs=4)

parser.add_option('-p', '--padding', dest='padding',
                  help='Extra margin of tiles to add around bounded area. Default value is %s (no extra tiles).' % repr(defaults['padding']),
                  type='int')

parser.add_option('-e', '--extension', dest='extension',
                  help='Optional file type for rendered tiles. Default value is %s.' % repr(defaults['extension']))

parser.add_option('-f', '--progress-file', dest='progressfile',
                  help="Optional JSON progress file that gets written on each iteration, so you don't have to pay close attention.")

parser.add_option('-q', action='store_false', dest='verbose',
                  help='Suppress chatty output, --progress-file works well with this.')

parser.add_option('-i', '--include-path', dest='include',
                  help="Add the following colon-separated list of paths to Python's include path (aka sys.path)")

parser.add_option('--tile-list', dest='tile_list',
                  help='Optional file of tile coordinates, a simple text list of Z/X/Y coordinates. Overrides --bbox and --padding.')

def generateCoordinates(ul, lr, zooms, padding):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Flood-fill coordinates based on two corners, a list of zooms and padding.
    """
    # start with a simple total of all the coordinates we will need.
    count = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)
        
        rows = lr_.row + 1 - ul_.row
        cols = lr_.column + 1 - ul_.column
        
        count += int(rows * cols)

    # now generate the actual coordinates.
    # offset starts at zero
    offset = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)

        for row in range(int(ul_.row), int(lr_.row + 1)):
            for column in range(int(ul_.column), int(lr_.column + 1)):
                coord = Coordinate(row, column, zoom)
                
                yield (offset, count, coord)
                
                offset += 1

def listCoordinates(filename):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Read coordinates from a file with one Z/X/Y coordinate per line.
    """
    coords = (line.strip().split('/') for line in open(filename, 'r'))
    coords = (map(int, (row, column, zoom)) for (zoom, column, row) in coords)
    coords = [Coordinate(*args) for args in coords]
    
    count = len(coords)
    
    for (offset, coord) in enumerate(coords):
        yield (offset, count, coord)

if __name__ == '__main__':
    options, zooms = parser.parse_args()

    if options.include:
        for p in options.include.split(':'):
            path.insert(0, p)

    from TileStache import parseConfigfile, getTile
    from TileStache.Core import KnownUnknown
    from TileStache.Caches import Disk, Multi
    
    from ModestMaps.Core import Coordinate
    from ModestMaps.Geo import Location

    try:
        if options.config is None:
            raise KnownUnknown('Missing required configuration (--config) parameter.')

        if options.layer is None:
            raise KnownUnknown('Missing required layer (--layer) parameter.')

        config = parseConfigfile(options.config)

        if options.layer in ('ALL', 'ALL LAYERS') and options.layer not in config.layers:
            # clean every layer in the config
            layers = config.layers.values()

        elif options.layer not in config.layers:
            raise KnownUnknown('"%s" is not a layer I know about. Here are some that I do know about: %s.' % (options.layer, ', '.join(sorted(config.layers.keys()))))

        else:
            # clean just one layer in the config
            layers = [config.layers[options.layer]]
        
        verbose = options.verbose
        extension = options.extension
        progressfile = options.progressfile
        
        lat1, lon1, lat2, lon2 = options.bbox
        south, west = min(lat1, lat2), min(lon1, lon2)
        north, east = max(lat1, lat2), max(lon1, lon2)

        northwest = Location(north, west)
        southeast = Location(south, east)

        for (i, zoom) in enumerate(zooms):
            if not zoom.isdigit():
                raise KnownUnknown('"%s" is not a valid numeric zoom level.' % zoom)

            zooms[i] = int(zoom)
        
        if options.padding < 0:
            raise KnownUnknown('A negative padding will not work.')

        padding = options.padding
        tile_list = options.tile_list

    except KnownUnknown, e:
        parser.error(str(e))

    for layer in layers:

        if tile_list:
            coordinates = listCoordinates(tile_list)
        else:
            ul = layer.projection.locationCoordinate(northwest)
            lr = layer.projection.locationCoordinate(southeast)
    
            coordinates = generateCoordinates(ul, lr, zooms, padding)
        
        for (offset, count, coord) in coordinates:
            path = '%s/%d/%d/%d.%s' % (layer.name(), coord.zoom, coord.column, coord.row, extension)
    
            progress = {"tile": path,
                        "offset": offset + 1,
                        "total": count}
    
            if options.verbose:
                print >> stderr, '%(offset)d of %(total)d...' % progress,
    
            try:
                mimetype, format = layer.getTypeByExtension(extension)
            except:
                #
                # It's not uncommon for layers to lack support for certain
                # extensions, so just don't attempt to remove a cached tile
                # for an unsupported format.
                #
                pass
            else:
                config.cache.remove(layer, coord, format)
    
            if options.verbose:
                print >> stderr, '%(tile)s' % progress
                    
            if progressfile:
                fp = open(progressfile, 'w')
                json_dump(progress, fp)
                fp.close()
